File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedArmor3.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.inventory.PlayerInventory;
import cn.nukkit.item.Item;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.recording.player.PlayerRecordNode;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 17:02.
 * All rights reserved
 */
public class PlayerUpdatedArmor3 implements PlayerUpdated {

    private Item item;

    public static PlayerUpdatedArmor3 of(Item item) {
        return new PlayerUpdatedArmor3(item);
    }

    @Override
    public int getUpdateTypeId() {
        return PlayerUpdated.TYPE_ARMOR_3;
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    @Override
    public void processTo(PlaybackNPC ghost) {
        if (ghost != null && ghost.getInventory() != null) {
            PlayerInventory inv = ghost.getInventory();
            inv.setArmorItem(3, item == null ? Item.get(Item.AIR) : item);
            inv.sendArmorContents(ghost.getViewers().values());
        }
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        node.setArmor3(item);
        return node;
    }

    public PlayerUpdatedArmor3(BinaryStream stream) {
        read(stream);
    }

    private PlayerUpdatedArmor3(Item item) {
        this.item = item;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof PlayerUpdatedArmor3 o)) return false;
        return (item.equals(o.item));
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putSlot(this.item);
    }

    @Override
    public void read(BinaryStream stream) {
        this.item = stream.getSlot();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedArmor3{" +
            "item=" + item +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedArmor3.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.inventory.ArmorInventory
import cn.nukkit.item.Item
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.player.PlayerRecordNode
import net.easecation.ghosty.serializer.NukkitItemSerializer
import net.easecation.ghosty.util.PersistenceBinaryStreamHelper

@Serializable
data class PlayerUpdatedArmor3(
    @Serializable(with = NukkitItemSerializer::class)
    val item: Item,
) : PlayerUpdated, UpdateWithState {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_ARMOR_3

    override fun processTo(ghost: PlaybackNPC) {
        val inv = ghost.armorInventory ?: return
        inv.setItem(ArmorInventory.SLOT_FEET, item)
        inv.sendContents(ghost.viewers.values)
    }

    override fun applyTo(node: PlayerRecordNode) = node.apply { armor3 = item }

    companion object {
        @JvmStatic
        fun of(item: Item) = PlayerUpdatedArmor3(item)

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedArmor3> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedArmor3> {
        override fun write(updated: PlayerUpdatedArmor3, stream: BinaryStream) {
            PersistenceBinaryStreamHelper.putItem(stream, updated.item)
        }

        override fun read(stream: BinaryStream): PlayerUpdatedArmor3 = read(stream, 3)

        override fun read(stream: BinaryStream, version: Int): PlayerUpdatedArmor3 {
            return when (version) {
                0 -> PlayerUpdatedArmor3(stream.slot)
                1, 2, 3 -> PlayerUpdatedArmor3(PersistenceBinaryStreamHelper.getItem(stream))
                else -> throw IllegalArgumentException("unknown format version: $version")
            }
        }
    }
}